src/ansys/acp/core/_recursive_copy.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (C) 2022 - 2024 ANSYS, Inc. and/or its affiliates.
+# SPDX-License-Identifier: MIT
+#
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import collections
+from collections.abc import Iterable
+
+import networkx as nx
+
+from ._dependency_graph import _build_dependency_graph, _WalkTreeOptions
+from ._tree_objects import LookUpTable1D, LookUpTable1DColumn, LookUpTable3D, LookUpTable3DColumn
+from ._tree_objects._grpc_helpers.linked_object_helpers import get_linked_paths
+from ._tree_objects.base import CreatableTreeObject, TreeObject
+from ._typing_helper import StrEnum
+
+__all__ = ["recursive_copy", "LinkedObjectHandling"]
+
+
+class LinkedObjectHandling(StrEnum):
+    """Defines options for handling linked objects when copying a tree of ACP objects."""
+
+    COPY = "copy"
+    KEEP = "keep"
+    DISCARD = "discard"
+
+
+def recursive_copy(
+    *,
+    source_objects: Iterable[CreatableTreeObject],
+    parent_mapping: dict[TreeObject, TreeObject],
+    linked_object_handling: LinkedObjectHandling | str = "copy",
```

Comment:
`keep` is the better default for me, since you typically stay within the model. I am aware of that this value is not valid if the copy is across models.

Reply:
Makes sense; we can add a check and provide an explicit error when `"keep"` is used.

It anyway also makes sense that all `source_objects` + the keys in `parent_mapping` belong to the same model, as well as all values in `parent_mapping`.